From 02845b214c00b7a01ad4b9dd15e31b991fe81514 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 27 May 2025 16:02:04 +0300
Subject: [PATCH] [JSW-704] jgs: address space update

With this patch added, pcie iccm address and
firmware memory address declarations.
Defined new macros to check certain address in
pre-defined address range or not, and macro to
convert 64bit address to 32bit arc address.

Change-Id: Ib331b7301dd5e6dab0f7cd50c4aa9dcead3d5d01
---

diff --git a/fs/jgs/address_space.h b/fs/jgs/address_space.h
index 3967616..3826b96 100644
--- a/fs/jgs/address_space.h
+++ b/fs/jgs/address_space.h
@@ -95,6 +95,8 @@
 #define IO_0_PID_ARC_DCCM_SIZE		(IO_0_CFG_FW_MEM_PID_ARC_DCCM_SECTION)
 #define IO_0_MISC_ARC_ICCM_SIZE		(IO_0_CFG_FW_MEM_MISC_ARC_ICCM_SECTION)
 #define IO_0_MISC_ARC_DCCM_SIZE		(IO_0_CFG_FW_MEM_MISC_ARC_DCCM_SECTION)
+#define IO_0_PCIE_ARC_ICCM_SIZE		(IO_0_CFG_FW_MEM_PCIE_ARC_ICCM_SECTION)
+#define IO_0_FW_MEM_SIZE		(IO_0_CFG_FW_MEM_SECTION)
 
 #define IO_0_RST_ARC_MEM_SIZE		(IO_0_RST_ARC_ICCM_SIZE + IO_0_RST_ARC_DCCM_SIZE)
 #define IO_0_MNG_ARC_MEM_SIZE		(IO_0_MNG_ARC_ICCM_SIZE + IO_0_MNG_ARC_DCCM_SIZE)
@@ -104,6 +106,7 @@
 
 #define IO_0_SP_SRAM_BASE		(mmIO_0_CFG_SP_RAM_BASE) /* TODO: SOL Update */
 #define IO_0_FW_MEM_BASE		(mmIO_0_CFG_FW_MEM_BASE)
+#define IO_0_PCIE_ARC_ICCM_BASE		(mmIO_0_CFG_FW_MEM_PCIE_ARC_ICCM_BASE)
 #define IO_0_PSOC_FW_SRAM_BASE		(mmIO_0_CFG_FW_MEM_FW_SRAM_BASE)
 #define IO_0_RST_ARC_ICCM_BASE		(mmIO_0_CFG_FW_MEM_RST_ARC_ICCM_BASE)
 #define IO_0_MNG_ARC_ICCM_BASE		(mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE)
@@ -112,6 +115,7 @@
 
 #define IO_0_SP_SRAM_BASE_FULL		(GET_FULL_ADDR(mmIO_0_CFG_SP_RAM_BASE)) /* TODO: SOL Update */
 #define IO_0_FW_MEM_BASE_FULL		(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_BASE))
+#define IO_0_PCIE_ARC_ICCM_BASE_FULL	(GET_FULL_ADDR(IO_0_PCIE_ARC_ICCM_BASE))
 #define IO_0_PSOC_FW_SRAM_BASE_FULL	(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_FW_SRAM_BASE))
 #define IO_0_RST_ARC_ICCM_BASE_FULL	(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_RST_ARC_ICCM_BASE))
 #define IO_0_MNG_ARC_ICCM_BASE_FULL	(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_MNG_ARC_ICCM_BASE))
@@ -119,6 +123,7 @@
 #define IO_0_MISC_ARC_ICCM_BASE_FULL	(GET_FULL_ADDR(mmIO_0_CFG_FW_MEM_MISC_ARC_ICCM_BASE))
 
 
+/* TODO: why this is not working for JGS? */
 #define CONV_64B_TO_ARC_FW_MEM(addr)	((addr & 0x3FFFFF) + mmIO_0_CFG_FW_MEM_BASE)
 
 
diff --git a/fs/jgs/soc.h b/fs/jgs/soc.h
index 00a1a9a..6348fc0 100644
--- a/fs/jgs/soc.h
+++ b/fs/jgs/soc.h
@@ -44,6 +44,12 @@
 
 #define GET_FULL_ADDR(REG_OFF)		((REG_OFF & MASK_28_BIT) + CFG_MEM_FULL_BASE)
 
+#define GET_ARC_ADDR(REG_OFF, REGION)   \
+    ((REG_OFF & MASK_28_BIT) + (REGION_BASE_ADDR(REGION)))
+
+#define CHECK_ADDR_IN_RANGE(ADDR, ADDR_O, BASE, BASE_O)	\
+    ((ADDR >= BASE) && ((ADDR + ADDR_O) <= (BASE + BASE_O)))
+
 /* This 2 macro's to convert ARC based 32B address to LBW MEM map address of 32B */
 #define CONV_ARC_REG_TO_GEN_MEM(addr)	(((addr) & MASK_28_BIT) + ((GEN_MEM_FULL_BASE) & MASK_32_BIT))
 #define CONV_ARC_REG_TO_CFG_MEM(addr)	(((addr) & MASK_28_BIT) + ((CFG_MEM_FULL_BASE) & MASK_32_BIT))
